File: AlgorithmsPackPythonSolutions/Kill Process.py
class Solution:
    def killProcess(self, pid: List[int], ppid: List[int], kill: int) -> List[int]:
        graph = {}
        for i in range(len(pid)):
            self.add_edge(graph, ppid[i], pid[i])
        
        return self.killer(graph, kill)
    
    def add_edge(self, graph, from_node, to_node):
        if from_node not in graph:
            graph[from_node] = []
        graph[from_node].append(to_node)
    
    def dfs(self, graph, kill, visited):
        visited.add(kill)
        if kill in graph:
            for node in graph[kill]:
                if node not in visited:
                    self.dfs(graph, node, visited)
    
    def killer(self, graph, kill):
        visited = set()
        self.dfs(graph, kill, visited)
        return list(visited)

File: AlgorithmsPackJavaSolutions/Kill Process.java
class Solution {
    public List<Integer> killProcess(List<Integer> pid, List<Integer> ppid, int kill) {
        Map<Integer, List<Integer>> graph = new HashMap<>();
        buildGraph(graph, pid, ppid);
        return killer(graph, kill);
    }

    private void buildGraph(Map<Integer, List<Integer>> graph, List<Integer> pid, List<Integer> ppid) {
        for (int i = 0; i < pid.size(); i++) {
            addEdge(graph, ppid.get(i), pid.get(i));
        }
    }

    private void addEdge(Map<Integer, List<Integer>> graph, int from, int to) {
        graph.computeIfAbsent(from, k -> new ArrayList<>()).add(to);
    }

    private void dfs(Map<Integer, List<Integer>> graph, int kill, Set<Integer> visited) {
        visited.add(kill);
        if (graph.containsKey(kill)) {
            for (int node : graph.get(kill)) {
                if (!visited.contains(node)) {
                    dfs(graph, node, visited);
                }
            }
        }
    }

    private List<Integer> killer(Map<Integer, List<Integer>> graph, int kill) {
        Set<Integer> visited = new HashSet<>();
        dfs(graph, kill, visited);
        return new ArrayList<>(visited);
    }
}
